staking/contracts/RariMineV3.sol

```diff
@@ -0,0 +1,140 @@
+import "@openzeppelin/contracts-upgradeable/utils/AddressUpgradeable.sol";
+import "@openzeppelin/contracts-upgradeable/token/ERC20/IERC20Upgradeable.sol";
+
+import "./libs/LibString.sol";
+import "./libs/LibAddress.sol";
+import "./libs/LibUint.sol";
+import "./libs/LibStakingMath.sol";
+import "./IRariMine.sol";
+import "./IStaking.sol";
+
+contract RariMineV3 is OwnableUpgradeable, IRariMine {
+    using SafeMathUpgradeable for uint;
+    using LibString for string;
+    using LibUint for uint;
+    using LibAddress for address;
+
+    IERC20Upgradeable public token;
+    address           public tokenOwner;
+    IStaking          public staking;
+
+    uint256 public claimFormulaClaim;
+    uint256 public claimCliffWeeks;
+    uint256 public claimSlopeWeeks;
+    uint256 constant CLAIM_FORMULA_DIVIDER = 10000; 
+
+    uint8   public constant VERSION        = 1;
+    
+    mapping(address => uint) public claimed;
+
+    event SetClaimFormulaClaim(uint256 indexed newClaimFormulaClaim);
+    event SetClaimCliffWeeks(uint256 indexed newClaimCliffWeeks);
+    event SetClaimSlopeWeeks(uint256 indexed newClaimSlopeWeeks);
+
+    function __RariMineV3_init(IERC20Upgradeable _token, 
+                                address _tokenOwner, 
+                                IStaking _staking, 
+                                uint256 _claimCliffWeeks, 
+                                uint256 _claimSlopeWeeks, 
+                                uint256 _claimFormulaClaim) external initializer {
+        __RariMineV3_init_unchained(_token, _tokenOwner, _staking, _claimCliffWeeks, _claimSlopeWeeks, _claimFormulaClaim);
+        __Ownable_init_unchained();
+        __Context_init_unchained();
+    }
+
+    function __RariMineV3_init_unchained(IERC20Upgradeable _token, 
+                                        address _tokenOwner, 
+                                        IStaking _staking, 
+                                        uint256 _claimCliffWeeks, 
+                                        uint256 _claimSlopeWeeks, 
+                                        uint256 _claimFormulaClaim) internal initializer {
+        token = _token;
+        tokenOwner = _tokenOwner;
+        staking = _staking;
+        claimCliffWeeks = _claimCliffWeeks;
+        claimSlopeWeeks = _claimSlopeWeeks;
+        claimFormulaClaim = _claimFormulaClaim;
+    }
+
+    function claim(Balance memory _balance, uint8 v, bytes32 r, bytes32 s) public {
+        require(prepareMessage(_balance, address(this)).recover(v, r, s) == owner(), "owner should sign balances");
+
+        address recipient = _balance.recipient;
+        if (_msgSender() == recipient) {
+            uint toClaim = _balance.value.sub(claimed[recipient], "nothing to claim");
+            claimed[recipient] = claimed[recipient].add(_balance.value);
+
+            // claim rari tokens
+            uint claimAmount = toClaim.mul(claimFormulaClaim).div(CLAIM_FORMULA_DIVIDER);
```

Comment:
By default this will be 0. Let's add if (to check if zefo, then don't do anything)

Reply:
done